### PR TITLE
Add support for watchOS apps

### PR DESCRIPTION
**Pull request by kwridan**
### Short description 📝

Adds support for watchOS apps.

### Solution 📦

Add support for watchOS / watchOS Extension products as well as appropriate dependency rules.

### Implementation 👩‍💻👨‍💻

- [x] Add product types
- [x] Add fixture & acceptance test
- [x] Update graph / link rules
- [x] Add embed watch build phase
- [x] Add watch build settings
- [x] Update changelog

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_app_with_watchapp2`
- Verify the generated project contains a Watch App & Watch App Extension
- Compare against a manually created project with a Watch App

### Notes

For follow up PRs:

- [ ] Update schemes to include a runnable watch apps (requires using `RemoteRunnable` launch actions)
- [ ] Add support for watch app containers (aka standalone watch apps)

### CODE REVIEW

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift [hunk 1]
```diff
@@ -286,4 +293,30 @@ final class BuildPhaseGenerator: BuildPhaseGenerating {
...
         }
     }
+
+    func generateEmbedWatchBuildPhase(path: AbsolutePath,
+                                      target: Target,
+                                      graph: Graphing,
+                                      pbxTarget: PBXTarget,
+                                      fileElements: ProjectFileElements,
+                                      pbxproj: PBXProj) throws {
+        let targetNode = graph.targetDependencies(path: path, name: target.name)
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]:
```suggestion
let targetDependencies = graph.targetDependencies(path: path, name: target.name)
```

Comment (kwridan) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]: 👍

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift [hunk 2]
```diff
@@ -286,4 +293,30 @@ final class BuildPhaseGenerator: BuildPhaseGenerating {
...
     }
+
+    func generateEmbedWatchBuildPhase(path: AbsolutePath,
+                                      target: Target,
+                                      graph: Graphing,
+                                      pbxTarget: PBXTarget,
+                                      fileElements: ProjectFileElements,
+                                      pbxproj: PBXProj) throws {
+        let targetDependencies = graph.targetDependencies(path: path, name: target.name)
+        let watchApps = targetDependencies.filter { $0.target.product == .watch2App }
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]: Interesting, I would have expected the watch app to have the iOS as a dependency, like we do with tests having the app as a dependency.

File: Sources/TuistGenerator/Generator/ConfigGenerator.swift [hunk 3]
```diff
@@ -153,6 +153,15 @@ final class ConfigGenerator: ConfigGenerating {
                                      target: Target,
                                      graph: Graphing,
                                      sourceRootPath: AbsolutePath) {
+        settings.merge(generalTargetDerrivedSettings(target: target, sourceRootPath: sourceRootPath)) { $1 }
+        settings.merge(testBundleTargetDerrivedSettings(target: target, graph: graph, sourceRootPath: sourceRootPath)) { $1 }
+        settings.merge(deploymentTargetDerrivedSettings(target: target)) { $1 }
+        settings.merge(watchTargetDerrivedSettings(target: target, graph: graph, sourceRootPath: sourceRootPath)) { $1 }
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/ConfigGenerator.swift]: Typo: watchTargetDerrivedSettings

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift (hunk 2 above)
Comment (kwridan) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]: Indeed this is an interesting one, using dependency paradigms for those can be thought of going either way depending how you look at it.

I opted to keep it consistent with how extensions are currently defined, they are fairly similar conceptually where they are companion products bundled with the main application - as such the application isn't complete without them.

```swift
 Target(name: "App",
        platform: .iOS,
        product: .app,
        bundleId: "io.tuist.App",
        infoPlist: "Support/App-Info.plist",
        sources: ["Sources/**"],
        dependencies: [

            // Extensions + Additional Bundled Products
            .target(name: "TodayExtension"),
            .target(name: "NotificationServiceExtension"),
            .target(name: "WatchApp"),
        ]),
```

For tests, its slightly different as often the test will actually depend on the app (`@testable import App`) additionally the test targets need to reference the host / target application either in their build settings or in their target attributes thus the dependency was specified in that direction.

File: Sources/TuistGenerator/Generator/ConfigGenerator.swift (hunk 3 above)
Comment (kwridan) [reply Sources/TuistGenerator/Generator/ConfigGenerator.swift]: 😅yikes, will fix those

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]: > I opted to keep it consistent with how extensions are currently defined, they are fairly similar conceptually where they are companion products bundled with the main application - as such the application isn't complete without them.

That's a fair reason 👍 and the example that you showed is very clear.

**Comment by kwridan** [comment]
>  If I'm not wrong, the bundle id of the extension has to be the same of the app suffixed by .watchextension. What do you think about adding a check for that? It took me a long time to figure out what was the issue and considering we know the convention, we can codify it into the Tuist linters.

👍 good idea - will update

**Pull request merged**